### PR TITLE
PS-296: mysqld crashing with subselects and long keys

### PR DESCRIPTION
**Pull request by dutow**
The heap-dynamic-rows patch limited the maximum key size of strings to 1000.
This worked fine in earlier versions, where the maximum key length was below 1000.
In 5.7, this was increased to 3072, which resulted in keys with 999 or more characters crashing in some operations.
The latest dynamic rows patch doesn't contain this limitation anymore, looks like it can be safely removed.

Note that this change doesn't fix the issue completely:
varchar/varbinary keys at least 3071 character long will still crash - in practice, this means a maximum length of 3071 or 3072 chararcters, as that's the current maximum.
This same bug is also present in mysql server.

### CODE REVIEW

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 1]
```diff
@@ -0,0 +1,36 @@
+
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test:3]: Add header block --echo # / --echo # PS-296: blah blah / --echo #

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+--disable_query_log
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: Why?

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+--disable_result_log
+create table a (id int unsigned not null auto_increment primary key);
+delimiter |;
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: One row too soon?

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+insert into a () values ();
+CREATE PROCEDURE dowhile()
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: Inconsistent SQL keyword case

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+BEGIN
+ DECLARE v1 INT DEFAULT 1;
+ while v1 < 15 do 
+insert into a select null from a;
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: Indentation

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+CREATE TABLE `big2` (`id` varbinary(999) NOT NULL, PRIMARY KEY (`id`)) ENGINE=InnoDB;
+insert into big1 select * from a;
+insert into big2 select * from a;
+set session optimizer_switch='duplicateweedout=off,materialization=on';
+--enable_result_log
+--enable_query_log
+
+select * from big1 where id in (select id from big2 ignore index (primary)) order by id asc limit 10;
+
+--disable_query_log
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: Why?

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+call dowhile();
+CREATE TABLE `big1` (`id` varbinary(999) NOT NULL,PRIMARY KEY (`id`)) ENGINE=InnoDB;
+CREATE TABLE `big2` (`id` varbinary(999) NOT NULL, PRIMARY KEY (`id`)) ENGINE=InnoDB;
+insert into big1 select * from a;
+insert into big2 select * from a;
+set session optimizer_switch='duplicateweedout=off,materialization=on';
+--enable_result_log
+--enable_query_log
+
+select * from big1 where id in (select id from big2 ignore index (primary)) order by id asc limit 10;
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: add explain select for the same too

Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: so this is due to a temp MEMORY table that stores VARBINARY. How do we end up getting one? Dynamic varchar patch did not enable them for the query optimizer use. Does the upstream use MEMORY table too?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 7 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: Yes, it's the same. And I don't see anything in the query code that would disable the memory engine for string joins - the heap engine's maximum key length is used both in upstream and our code for this join, even for longer strings.

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 6 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: I thought that it would be good if the result file only focused on the statements which are important for the crash.

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 2 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: I'll remove both this and the later disabling instructions - but keeping the result log disabled is probably a good idea here, because that stored procedure inserts a few thousands records.

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 7 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: ops, I deleted this comment, as it wasn't true - local changes resulted in some incorrect results

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 8]
```diff
@@ -0,0 +1,31 @@
+-- echo PS-296: inconsistent maximum key lenghts caused a crash in the query optimizer
+
+--disable_result_log
```
Comment (percona-ysorokin) [inline mysql-test/t/percona_heap_bug_ps296.test]: Add `--source include/have_innodb.inc`

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+insert into a () values ();
+delimiter |;
+create procedure dowhile()
+begin
+  declare v1 int default 1;
+  while v1 < 15 do 
+  insert into a select null from a;
+  set v1 = v1 + 1;
+  end while;
+  end;
```
Comment (percona-ysorokin) [inline mysql-test/t/percona_heap_bug_ps296.test]: indentation

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+--disable_result_log
+create table a (id int unsigned not null auto_increment primary key);
+insert into a () values ();
+delimiter |;
+create procedure dowhile()
+begin
+  declare v1 int default 1;
+  while v1 < 15 do 
+  insert into a select null from a;
```
Comment (percona-ysorokin) [inline mysql-test/t/percona_heap_bug_ps296.test]: indentation for the inner `while` block

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 11]
```diff
@@ -0,0 +1,31 @@
+-- echo PS-296: inconsistent maximum key lenghts caused a crash in the query optimizer
+
+--disable_result_log
+create table a (id int unsigned not null auto_increment primary key);
```
Comment (percona-ysorokin) [inline mysql-test/t/percona_heap_bug_ps296.test]: Please, also consider uppercase for SQL keywords - this will make the testcase more readable.

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 8 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: Done

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 11 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: Done

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 10 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: Done

File: mysql-test/t/percona_heap_bug_ps296.test (hunk 9 above)
Comment (dutow) [reply mysql-test/t/percona_heap_bug_ps296.test]: Done

**Comment by dutow** [comment]
PR updated.

File: mysql-test/t/percona_heap_bug_ps296.test [hunk 12]
```diff
@@ -0,0 +1,32 @@
+-- echo PS-296: inconsistent maximum key lenghts caused a crash in the query optimizer
+--source include/have_innodb.inc
+
+--disable_result_log
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_bug_ps296.test]: disable result log only before the loop calls

File: include/heap.h [hunk 13]
```diff
@@ -42,7 +42,7 @@ extern "C" {
 #define HP_MAX_KEY                  MAX_INDEXES         /* Max allowed keys */
 #endif
 
-#define HP_MAX_KEY_LENGTH           1000            /* Max length in bytes */
+#define HP_MAX_KEY_LENGTH           MAX_KEY_LENGTH      /* Max length in bytes */
```
Comment (laurynas-biveinis) [inline include/heap.h:45]: I'm sorry for not commenting this earlier, but this is a functionality change for MEMORY SE, thus needs at least the most basic test: CREATE TABLE w/ index lengths > 1000 < new maximum, check insert, reads using that index, other reads. Check that CREATE TABLE w/ index lengths > newmax fails OK. Check that composite key with key parts > 1000 < newmax, sum < newmax work, sum > newmax fail.

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: include/heap.h (hunk 13 above)
Comment (dutow) [reply include/heap.h:45]: Uh, we have another issue there, and I was able to reproduce this in upstream, both 5.7 and 8.0:

varchar and varbinary ordering fails in memory tables if the key is longer than 1022 characters. (as the length adds 2 more bytes, as would be my guess)

As this is present in the upstream too, and it's a different bug than the crash, shouldn't we handle it as a different issue?

Comment (dutow) [reply include/heap.h:45]: I figured this out, and added a test (with a single key for now). The issue is that the max_sort_length variable defaults to 1024, while the max key length is 3072.

For now, I changed the max_sort_length value for the test sessions, but maybe it would be a good idea to change its default to be 3072 too - otherwise a query's result will depend on the sorting algorithm used.

**Review by laurynas-biveinis — APPROVED** [review]

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Comment by dutow** [comment]
Made two small adjustments:
* rerecorded a test result - it had warnings about too long keys, which now aren't there (these warnings were ignored by the test)
* forgot to drop a table in the test for the bug

jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1647/

**Review by laurynas-biveinis — APPROVED** [review]
